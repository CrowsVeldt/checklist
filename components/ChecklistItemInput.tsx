import { useEffect, useRef, useState } from "react";
import { Pressable, StyleSheet, TextInput, View } from "react-native";
import RequiredButton from "./RequiredButton";
import DeleteEntryButton from "./DeleteEntryButton";

export default function ChecklistItemInput({
  id,
  initialTitle,
  initialRequired,
  child,
  onEntryChange,
  remove,
  first,
  last,
}: {
  id: string;
  initialTitle: string;
  child?: boolean;
  initialRequired?: boolean;
  onEntryChange: any;
  remove: () => void;
  first?: boolean;
  last?: boolean;
}) {
  const [title, setTitle] = useState<string>(
    initialTitle != null ? initialTitle : "Entry Title"
  );
  const [required, setRequired] = useState<boolean>(
    initialRequired != null ? initialRequired : false
  );
  const [selection, setSelection] = useState<
    { start: number; end: number } | undefined
  >(undefined);

  const inputRef = useRef<TextInput>(null);

  useEffect(() => {
    onEntryChange(id, title, required, child);
  }, [title, required]);

  return (
    <Pressable
      style={[
        styles.inputWrapper,
        first ? styles.firstItem : {},
        last ? styles.lastItem : {},
      ]}
      onPress={() => {
        if (inputRef.current) {
          inputRef.current.focus();
          // Set cursor position to the end of the text when inputWrapper is pressed
          setSelection({ start: title.length, end: title.length });
        }
      }}
    >
      <TextInput
        value={title}
        onChangeText={setTitle}
        ref={inputRef}
        selection={selection}
        onFocus={() => {
          // set cursor position to undefined so that the user can change it
          setSelection(undefined);
        }}
      />
      <View style={styles.buttonWrapper}>
        <RequiredButton initialState={required} changeState={setRequired} />
        <DeleteEntryButton remove={remove} />
      </View>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  inputWrapper: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    borderWidth: 1,
    borderColor: "black",
    height: 70,
    marginBottom: 3,
  },
  firstItem: {
    borderTopStartRadius: 10,
    borderTopEndRadius: 10,
  },
  lastItem: {
    borderBottomStartRadius: 10,
    borderBottomEndRadius: 10,
  },
  buttonWrapper: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    height: "100%",
  },
});
