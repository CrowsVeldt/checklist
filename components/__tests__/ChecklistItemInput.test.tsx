import { render } from "@testing-library/react-native";
import ChecklistItemInput from "../ChecklistItemInput";

describe("<ChecklistItemInput />", () => {
  test("ChecklistItemInput renders correctly", () => {
    const tree = render(
      <ChecklistItemInput
        id="abc"
        initialTitle="title"
        onEntryChange={() => true}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
