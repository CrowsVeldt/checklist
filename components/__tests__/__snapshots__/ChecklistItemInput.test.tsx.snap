// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChecklistItemInput /> ChecklistItemInput renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderColor": "black",
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 70,
        "justifyContent": "space-between",
        "marginBottom": 3,
      },
      {},
      {},
    ]
  }
>
  <TextInput
    onChangeText={[Function]}
    value="title"
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": "100%",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderStartColor": "black",
          "borderStartWidth": 1,
          "height": "100%",
          "justifyContent": "center",
          "width": 50,
        }
      }
    >
      <Text />
      <View
        style={
          {
            "borderColor": "gray",
            "borderWidth": 1,
            "position": "absolute",
            "transform": [
              {
                "rotate": "45deg",
              },
            ],
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderStartColor": "black",
          "borderStartWidth": 1,
          "height": "100%",
          "justifyContent": "center",
          "width": 50,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
